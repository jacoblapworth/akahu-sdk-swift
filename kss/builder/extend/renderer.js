const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
