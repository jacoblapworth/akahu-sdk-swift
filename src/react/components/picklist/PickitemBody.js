import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import XUICheckbox from '../checkbox/XUICheckbox';

const NOOP = () => {};

/**
 * INTERNAL USE ONLY
 *
 * This presentation component is used internally in the XUI library to display the
 * contents of a Pickitem.
 *
 * @param {Object} props
 */
const PickitemBody = ({ onClick, onKeyDown, shouldTruncate, onMouseOver, isSelected, href, isMultiselect, children, checkboxClassName, target, qaHook }) => {
	if (isMultiselect) {
		return (
			<div className="xui-pickitem--body" onClick={onClick} onKeyDown={onKeyDown} onMouseOver={onMouseOver} data-automationid={qaHook}>
				<XUICheckbox onChange={NOOP} isChecked={isSelected} svgClassName="xui-pickitem--input" className={checkboxClassName} tabIndex={-1} qaHook={qaHook && `${qaHook}--checkbox`}>
					<span className="xui-pickitem--multiselect-label" data-automationid={`${qaHook}--label`}>{children}</span>
				</XUICheckbox>
			</div>
		);
	}
	const rel = target ? "noopener noreferrer" : null;
	const childProps = {
		className: 'xui-pickitem--body',
		tabIndex: '-1',
		onClick,
		onKeyDown,
		onMouseOver,
		rel,
	};

	const textClassName = cn('xui-pickitem--text', {
		'xui-text-truncated': shouldTruncate
	});

	const text = <span className={textClassName}>{children}</span>;
	return href ? (
		<a href={href} target={target} data-automationid={qaHook} {...childProps}>{text}</a>
	) : (
		<button type="button" data-automationid={qaHook} {...childProps}>{text}</button>
	);
};

PickitemBody.propTypes = {
	children: PropTypes.node,
	isSelected: PropTypes.bool,
	isMultiselect: PropTypes.bool,
	href: PropTypes.string,
	checkboxClassName: PropTypes.string,
	onClick: PropTypes.func,
	onKeyDown: PropTypes.func,
	onMouseOver: PropTypes.func,
	target: PropTypes.string,
	shouldTruncate: PropTypes.bool,
	qaHook: PropTypes.string
};

export default PickitemBody;
