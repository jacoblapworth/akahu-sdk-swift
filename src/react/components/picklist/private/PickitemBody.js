import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { pickitemClassName, itemBodyClassName, itemTextClassName } from './constants';
import { XUIButtonCaret } from '../../../button';

/**
 * INTERNAL USE ONLY
 *
 * This presentation component is used internally in the XUI library to display the
 * contents of a Pickitem, when not multiselect.
 *
 * @param {Object} props
 */
const PickitemBody = ({
  onClick,
  onKeyDown,
  shouldTruncate,
  onMouseOver,
  onBlur,
  onFocus,
  href,
  children,
  target,
  qaHook,
  tabIndex,
  primaryElement,
  secondaryElement,
  pinnedElement,
  leftElement,
  rightElement,
  headingElement,
  showButtonCaret,
}) => {
  const rel = target ? 'noopener noreferrer' : null;
  const childProps = {
    className: itemBodyClassName,
    onClick,
    onKeyDown,
    onMouseOver,
    onBlur,
    onFocus,
    rel,
    tabIndex,
  };

  const textClassName = cn(
    itemTextClassName,
    shouldTruncate && `${pickitemClassName}-text-truncated`,
  );

  const Tag = href ? 'a' : 'button';
  const elementSettings = href ? { ...{ href, target } } : { type: 'button' };

  const mainContent =
    (children || primaryElement) && shouldTruncate ? (
      <span className={`${pickitemClassName}-text-truncated`}>
        {primaryElement}
        {children}
      </span>
    ) : (
      <Fragment>
        {primaryElement}
        {children}
      </Fragment>
    );

  return (
    <Tag {...elementSettings} data-automationid={qaHook} {...childProps}>
      {leftElement}
      <span className={textClassName}>
        {headingElement}
        {mainContent}
        {secondaryElement}
      </span>
      {pinnedElement}
      {rightElement}
      {showButtonCaret && <XUIButtonCaret />}
    </Tag>
  );
};

PickitemBody.propTypes = {
  children: PropTypes.node,
  href: PropTypes.string,
  checkboxClassName: PropTypes.string,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  onMouseOver: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  target: PropTypes.string,
  shouldTruncate: PropTypes.bool,
  qaHook: PropTypes.string,
  tabIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Standard text */
  primaryElement: PropTypes.node,
  /** Less important text to appear beside primary. */
  secondaryElement: PropTypes.node,
  /** Less important text to appear pinned at the right. */
  pinnedElement: PropTypes.node,
  /** Content to be added to the left of the pickitem. */
  leftElement: PropTypes.node,
  /** Content to be added to the right of the pickitem. */
  rightElement: PropTypes.node,
  headingElement: PropTypes.node,
  /** Show button caret. Used in TabDropDown */
  showButtonCaret: PropTypes.bool,
};

export default PickitemBody;
