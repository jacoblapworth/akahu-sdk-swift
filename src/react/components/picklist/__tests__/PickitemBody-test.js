import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { axe, toHaveNoViolations } from 'jest-axe';
import renderer from 'react-test-renderer';
import PickitemBody from '../private/PickitemBody';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

describe('<PickitemBody />', () => {
  it('creates a button by default', () => {
    const button = renderer.create(<PickitemBody>Item</PickitemBody>);
    expect(button).toMatchSnapshot();
  });

  it('adds truncation classes', () => {
    const truncation = renderer.create(<PickitemBody shouldTruncate>Item</PickitemBody>);
    expect(truncation).toMatchSnapshot();
  });

  it('creates an `a` tag when href is provided', () => {
    const anchor = renderer.create(<PickitemBody href="https://xero.com">Item</PickitemBody>);
    expect(anchor).toMatchSnapshot();
  });

  it('should pass accessibility testing', async () => {
    const wrapper = mount(<PickitemBody>Item</PickitemBody>);
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });
});
