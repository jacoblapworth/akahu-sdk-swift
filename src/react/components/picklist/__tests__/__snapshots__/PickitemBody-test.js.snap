// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PickitemBody /> adds truncation classes 1`] = `
<button
  className="xui-pickitem--body"
  data-automationid={undefined}
  onClick={undefined}
  onKeyDown={undefined}
  onMouseOver={undefined}
  rel={null}
  tabIndex={undefined}
  type="button"
>
  <span
    className="xui-pickitem--text xui-pickitem-text-truncated"
  >
    <span
      className="xui-pickitem-text-truncated"
    >
      Item
    </span>
  </span>
</button>
`;

exports[`<PickitemBody /> creates a button by default 1`] = `
<button
  className="xui-pickitem--body"
  data-automationid={undefined}
  onClick={undefined}
  onKeyDown={undefined}
  onMouseOver={undefined}
  rel={null}
  tabIndex={undefined}
  type="button"
>
  <span
    className="xui-pickitem--text"
  >
    Item
  </span>
</button>
`;

exports[`<PickitemBody /> creates an \`a\` tag when href is provided 1`] = `
<a
  className="xui-pickitem--body"
  data-automationid={undefined}
  href="https://xero.com"
  onClick={undefined}
  onKeyDown={undefined}
  onMouseOver={undefined}
  rel={null}
  tabIndex={undefined}
  target={undefined}
>
  <span
    className="xui-pickitem--text"
  >
    Item
  </span>
</a>
`;
